cmd/controller/main.go

```diff
@@ -211,14 +216,44 @@ func main() {
 		}
 	}
 
-	startUtilHttpServerOrDie(kubeClient, kubeconfigProvider, stopCh)
-	startControllerViaLeaderElectionOrDie(runOptions)
+	ctx, ctxDone := context.WithCancel(context.Background())
+	runOptions.ctx = ctx
+	var g run.Group
+	{
+		g.Add(func() error {
+			return startUtilHttpServer(kubeClient, kubeconfigProvider, ctx)
+		}, func(err error) {
+			ctxDone()
```

Comment:
Instead of `ctxDone()` run `server.Shutdown()` on the server you created above in the same group.